### ISSUE TITLE
Bump Rust version to 1.56.0

### ISSUE THREAD

**Issue by TonalidadeHidrica** [rust-lang-ja/ac-library-rs#118]
Currently, the MSRV (Minimum Supported Rust Version) is essentially 1.42.0, which is intednded to work in AtCoder.  
Now that AtCoder is updating the version, it may be a good idea to bump supported version, so that we can now use 2021 Edition and various new features.

However we also need to consider other websites (CodeForces, yukicoder, ...) too.  I'm going to survey the Rust versions used in other platforms and post them as a succeeding comment.

**Comment by TonalidadeHidrica** [comment]
Here is a collection of "famous" online judge sites, accompanied by available rust versions.  (The choice is very subjective.  Feel free to suggest other websites that are somewhat famous.  I also tried to include as much Japanese judges as possible.)

- AtCoder: [1.42.0](https://atcoder.jp/contests/abc296/rules#:~:text=ruby%20./Main.rb%27-,Rust%20(1.42.0),-bash%20%2Dc%20%27cp) -> [1.68.1](https://docs.google.com/spreadsheets/d/1HXyOXt5bKwhKWXruzUvfMFHQtBxfZQ0047W7VVObnXI/edit#gid=408033513&range=H2)
- CodeForces: [1.66.0 (2021)](https://codeforces.com/contest/1798/submit) (could not find a blog post)
- TopCoder: [not supported](https://community.topcoder.com/tc?module=Static&d1=help&d2=codingWindow#:~:text=TopCoder%20currently%20allows%20coders%20to%20utilize%20Java%2C%20C%2B%2B%2C%20C%23%2C%20or%20Visual%20Basic%20.Net%20as%20a%20competition%20programming%20language.)
- CodeChef: [1.56](https://www.codechef.com/wiki/list-compilers#:~:text=rust-,rust%201.56,-Scala)
- oj.uz: not supported?
- LeetCode: [1.58.2](https://support.leetcode.com/hc/en-us/articles/360011833974-What-are-the-environments-for-the-programming-languages-#:~:text=Rust,1.58.2)
- HackerRank: [1.59.0](https://candidatesupport.hackerrank.com/hc/en-us/articles/4402913877523-Execution-Environment-and-Samples#:~:text=RUST,Version%201.59.0)
- Sphere Online Judge (SPOJ): [1.56](https://www.spoj.com/ranks/languages/#:~:text=RUST-,Rust%20(rust%201.56),-1120)
- CS Academy: [1.61.0](https://csacademy.com/about/environment/#:~:text=Rust-,rustc%201.61.0,-Kotlin)
- yukicoder: [1.59.0 (2018)](https://yukicoder.me/help/environments#:~:text=Rust,rs%20%2Do%20Main)
- PKU (Peking University) Online Judge (POJ): [not supported](http://poj.org/page?id=1000#:~:text=The%20judge%20system%20currently%20services%20solutions%20written%20in%20C%2C%20C%2B%2B%2C%20Pascal%2C%20Java%20and%20Fortran)
- Library Checker: [1.60.0](https://github.com/yosupo06/library-checker-judge/blob/master/langs/Dockerfile.RUST#LL5)
- Google Code Jam (GCJ): [1.41.0](https://codingcompetitions.withgoogle.com/codejam/faq#:~:text=Rust%3A,1.41.1) (will be closed on 2023-07-01 :cry:)
- Aizu Online Judge (AOJ): [1.42.1 ?](https://twitter.com/cm_ayf/status/1576584673991741441) (could not find an official source)
- Hamako Online Judge (HOJ): [not supported](https://hoj.hamako-ths.ed.jp/onlinejudge/info/2)
- Kotamanegi Online Judge: [not supported](https://kotamanegi.com/#:~:text=%E7%8F%BE%E5%9C%A8%E3%81%AFC%2B%2B14%2CJava8.0%2CPython2%2CPython3%2Cbash%E3%81%AB%E5%AF%BE%E5%BF%9C%E3%81%97%E3%81%A6%E3%81%84%E3%81%BE%E3%81%99%E3%80%82)
- Kagamiz Contest System (KCS): down
- CafeCoder: down
- MojaCoder: [1.43.0](https://mojacoder.app/users/Makutamoto/contests/test-contest-1/tasks/1) (could not find an official source)
- paiza: [1.59.0](https://paiza.jp/guide/language#:~:text=Rust,rustc%201.59.0)
- CodInGame:  [Shuold be written here, but it seems temporarily down](https://www.codingame.com/playgrounds/40701/help-center/languages-versions)

FYI: Rust 2021 Edition was stablized since [Rust 1.56.0](rust-lang/rust#88010).

---

Basically, I think we may bump MSRV to Rust 1.59.0, adopting Rust 2021.  Apart from GCJ, the only websites that will be no longer supported are AOJ and MojaCoder.   Do you think it is OK, or shuold we stick to 1.42-ish versions?

**Comment by TonalidadeHidrica** [comment]
My bad, I shuold have checked #83 first.  Anyway, this issue suggest a completely different direction than #83 , and I want other users' opinions.

**Comment by TonalidadeHidrica** [comment]
One possible downside of upgrading the edition to 2021 is that it may no longer compile when "expanded."

**Comment by mizar** [comment]
> * yukicoder: [1.59.0 (2018)](https://yukicoder.me/help/environments#:~:text=Rust,rs%20%2Do%20Main)

yukicoder Bumped [1.68.1 (2018)](https://yukicoder.me/help/environments#:~:text=Rust,rs%20%2Do%20Main).

**Comment by togatoga** [comment]
@TonalidadeHidrica 

Thank you for brining this topic. I think it's enough to stick with AtCoder platform because `ac-library-rs is a rust port of AtCoder Library`.

**Comment by koba-e964** [comment]
Universal Cup ([contest.ucup.ac](https://contest.ucup.ac/)) seems to support Rust 1.66.0.
(evidence: the following code
```Rust
fn main() {
// 1.66.0
let x = Some((1, "hi"));
let y = None::<(u8, u32)>;

assert_eq!(x.unzip(), (Some(1), Some("hi")));
assert_eq!(y.unzip(), (None, None));
// 1.67.0
assert_eq!(5i32.checked_ilog(5), Some(1));
}
```
emits a compilation error: https://contest.ucup.ac/submission/225227)

**Comment by koba-e964** [comment]
Personal opinion: we want to support AtCoder (1.70.0) and Codeforces (1.72.0) because:
- AtCoder is the primary objective of this platform
- Codeforces wields the biggest stick in competitive programming

All other platforms are insignificant compared to these two, so we can update to 1.70.0.

**Comment by koba-e964** [comment]
Closing because https://github.com/rust-lang-ja/ac-library-rs/pull/137 was merged.
